Error in stats subcommand

`gurnel stats` errors out under certain conditions. Steps to reproduce:

```
mkdir foo
gurnel start
# exit without writing any words
gurnel stats
100.00% of days journaled since Oct 18 2019
Total word count: 0
Average word count: 0.0

Top 100 unusually frequent words:
panic: runtime error: slice bounds out of range [:100] with capacity 0

goroutine 1 [running]:
github.com/mikeraimondi/gurnel/internal/gurnel.(*statsCmd).Run(0x19e91a0, 0x14b7da0, 0xc000086000, 0x14b7dc0, 0xc00000e010, 0xc00007e030, 0x0, 0x0, 0xc0000e6080, 0x0, ...)
        /__w/gurnel/gurnel/internal/gurnel/stats.go:149 +0x1a6a
github.com/mikeraimondi/gurnel/internal/gurnel.Do(0xc0000dff20, 0x103aa27)
        /__w/gurnel/gurnel/internal/gurnel/command.go:58 +0x571
main.main()
        /__w/gurnel/gurnel/cmd/gurnel/main.go:11 +0x26
```

--- a/internal/gurnel/beeminder.go
+++ b/internal/gurnel/beeminder.go
@@ -32,7 +32,11 @@ func newBeeminderClient(user string, token []byte) (*beeminderClient, error) {
 	}, nil
 }
 
-func (client *beeminderClient) postDatapoint(goal string, count int) error {
+func (client *beeminderClient) postDatapoint(
+	goal string,
+	count int,
+	t time.Time,
+) error {
 	if goal == "" {
 		return fmt.Errorf("goal must not be blank")
 	}
@@ -50,7 +54,7 @@ func (client *beeminderClient) postDatapoint(goal string, count int) error {
 	v := url.Values{}
 	v.Set("auth_token", string(client.Token))
 	v.Set("value", strconv.Itoa(count))
-	v.Set("comment", "via Gurnel at "+time.Now().Format("15:04:05 MST"))
+	v.Set("comment", "via Gurnel at "+t.Format("15:04:05 MST"))
 
 	resp, err := client.c.PostForm(postURL.String(), v)
 	if err != nil {

--- a/internal/gurnel/beeminder_test.go
+++ b/internal/gurnel/beeminder_test.go
@@ -9,6 +9,8 @@ import (
 	"strconv"
 	"strings"
 	"testing"
+
+	"github.com/mikeraimondi/gurnel/internal/test"
 )
 
 func TestBeeminderPostServerErrors(t *testing.T) {
@@ -51,9 +53,10 @@ func TestBeeminderPostServerErrors(t *testing.T) {
 				serverURL: server.URL,
 			}
 
+			now := (&test.FixedClock{}).Now()
 			result := make(chan error)
 			go func() {
-				result <- client.postDatapoint("foo", 1)
+				result <- client.postDatapoint("foo", 1, now)
 			}()
 			err := <-result
 
@@ -139,9 +142,10 @@ func TestBeeminderPostParameters(t *testing.T) {
 				serverURL: server.URL,
 			}
 
+			now := (&test.FixedClock{}).Now()
 			result := make(chan error)
 			go func() {
-				result <- client.postDatapoint(tt.goal, tt.count)
+				result <- client.postDatapoint(tt.goal, tt.count, now)
 			}()
 			err := <-result
 			if !tt.valid {
@@ -246,7 +250,8 @@ func TestBeeminderClient(t *testing.T) {
 				c:         http.Client{Transport: tt.transport},
 			}
 
-			err := client.postDatapoint("test", 10)
+			now := (&test.FixedClock{}).Now()
+			err := client.postDatapoint("test", 10, now)
 			if tt.expectedErr == "" {
 				if err != nil {
 					t.Fatalf("expected no error. got %q", err)

--- a/internal/gurnel/command.go
+++ b/internal/gurnel/command.go
@@ -8,6 +8,7 @@ import (
 	"os"
 	"strings"
 	"text/template"
+	"time"
 )
 
 type subcommand interface {
@@ -18,12 +19,17 @@ type subcommand interface {
 	Flag() flag.FlagSet
 }
 
+type defaultClock struct{}
+
+func (c *defaultClock) Now() time.Time { return time.Now() }
+
 // Do executes the program
 func Do() error {
 	var conf config
 	if err := conf.load("gurnel", "gurnel.json"); err != nil {
 		return fmt.Errorf("loading config: %w", err)
 	}
+	conf.clock = &defaultClock{}
 
 	flag.Usage = func() {
 		printUsage(conf.subcommands, os.Stderr)

--- a/internal/gurnel/config.go
+++ b/internal/gurnel/config.go
@@ -6,10 +6,15 @@ import (
 	"io/ioutil"
 	"os"
 	"path/filepath"
+	"time"
 )
 
 type dirProvider interface {
 	getConfigDir() (string, error)
+}
+
+type clock interface {
+	Now() time.Time
 }
 
 type config struct {
@@ -21,6 +26,7 @@ type config struct {
 	Editor             string
 	dp                 dirProvider
 	subcommands        []subcommand
+	clock              clock
 }
 
 type defaultDirProvider struct{}

--- a/internal/gurnel/journalentry.go
+++ b/internal/gurnel/journalentry.go
@@ -32,7 +32,7 @@ type Entry struct {
 
 // NewEntry reads the directory named by dir and either returns an existing
 // Entry in that directory, or creates a new one if none exist.
-func NewEntry(dir string) (*Entry, error) {
+func NewEntry(dir string, t time.Time) (*Entry, error) {
 	info, err := os.Stat(dir)
 	if err != nil {
 		return nil, err
@@ -40,10 +40,10 @@ func NewEntry(dir string) (*Entry, error) {
 	if !info.IsDir() {
 		return nil, errors.New("must be a directory")
 	}
-	p := &Entry{Path: dir + string(filepath.Separator) + time.Now().Format(entryFormat)}
+	p := &Entry{Path: dir + string(filepath.Separator) + t.Format(entryFormat)}
 	_, err = os.Stat(p.Path)
 	if os.IsNotExist(err) {
-		p.ModTime = time.Now()
+		p.ModTime = t
 		err = p.Save()
 	} else if err == nil {
 		_, err = p.Load()

--- a/internal/gurnel/start.go
+++ b/internal/gurnel/start.go
@@ -34,7 +34,7 @@ func (*startCmd) Run(r io.Reader, w io.Writer, args []string, conf *config) erro
 	if err != nil {
 		return errors.New("evaluating symlinks " + err.Error())
 	}
-	p, err := NewEntry(wd)
+	p, err := NewEntry(wd, conf.clock.Now())
 	if err != nil {
 		return err
 	}
@@ -46,7 +46,7 @@ func (*startCmd) Run(r io.Reader, w io.Writer, args []string, conf *config) erro
 	}
 	editCmd := strings.Split(editor, " ")
 	editCmd = append(editCmd, p.Path)
-	startTime := time.Now()
+	startTime := conf.clock.Now()
 	cmd := exec.Command(editCmd[0], editCmd[1:]...)
 	cmd.Stdin = r
 	cmd.Stdout = w
@@ -113,7 +113,7 @@ func (*startCmd) Run(r io.Reader, w io.Writer, args []string, conf *config) erro
 			if err != nil {
 				return fmt.Errorf("setting up client: %w", err)
 			}
-			err = client.postDatapoint(conf.BeeminderGoal, wordCount)
+			err = client.postDatapoint(conf.BeeminderGoal, wordCount, conf.clock.Now())
 			if err != nil {
 				return fmt.Errorf("posting to Beeminder: %w", err)
 			}

--- a/internal/gurnel/start_test.go
+++ b/internal/gurnel/start_test.go
@@ -9,6 +9,8 @@ import (
 	"runtime"
 	"strings"
 	"testing"
+
+	"github.com/mikeraimondi/gurnel/internal/test"
 )
 
 type testReader struct {
@@ -61,13 +63,11 @@ func TestStart(t *testing.T) {
 			_, filename, _, _ := runtime.Caller(0)
 			dir := filepath.Dir(filepath.Dir(filepath.Dir(filename)))
 			tC.conf.Editor = filepath.Join(dir, "test", "no_op_editor.sh")
-			dir, err := ioutil.TempDir("", "gurnel_test")
-			if err != nil {
-				t.Fatal(err)
-			}
-			if err = os.Chdir(dir); err != nil {
-				t.Fatal(err)
-			}
+			tC.conf.clock = &test.FixedClock{}
+
+			dir, cleanup := test.SetupTestDir(t)
+			defer cleanup()
+
 			cmd := startCmd{}
 			inReader := testReader{
 				t:     t,
@@ -91,17 +91,10 @@ func TestStart(t *testing.T) {
 				}
 				files, _ = ioutil.ReadDir(dir)
 			}
-			f, err := os.OpenFile(filepath.Join(dir, file.Name()), os.O_RDWR|os.O_APPEND|os.O_SYNC, 0600)
-			if err != nil {
-				t.Fatalf("error opening file: %s", err)
-			}
-			defer f.Close()
 
-			if _, err = f.WriteString(tC.input); err != nil {
-				t.Fatalf("error writing file: %s", err)
-			}
+			defer test.WriteFile(t, filepath.Join(dir, file.Name()), tC.input)()
 
-			err = <-errC
+			err := <-errC
 			if tC.err == "" {
 				if err != nil {
 					t.Fatalf("expected no error. got %s", err)
@@ -114,7 +107,7 @@ func TestStart(t *testing.T) {
 
 			for _, expectedOut := range tC.out {
 				if !strings.Contains(strings.ToLower(out.String()), strings.ToLower(expectedOut)) {
-					t.Fatalf("expected output containing %s. got %q", tC.out, out.String())
+					t.Fatalf("expected output containing %s. got %q", expectedOut, out.String())
 				}
 			}
 		})

--- a/internal/gurnel/stats.go
+++ b/internal/gurnel/stats.go
@@ -87,7 +87,7 @@ func (*statsCmd) Run(_ io.Reader, w io.Writer, args []string, conf *config) erro
 	}()
 	var entryCount float64
 	wordMap := make(map[string]uint64)
-	t := time.Now()
+	t := conf.clock.Now()
 	minDate := t
 	for r := range c {
 		if r.err != nil {
@@ -105,59 +105,64 @@ func (*statsCmd) Run(_ io.Reader, w io.Writer, args []string, conf *config) erro
 	if err := <-errc; err != nil {
 		return err
 	}
-	if entryCount > 0 {
-		percent := entryCount / math.Floor(t.Sub(minDate).Hours()/24)
-		const outFormat = "Jan 2 2006"
-		fmt.Fprintf(w, "%.2f%% of days journaled since %v\n", percent*100, minDate.Format(outFormat))
-		var wordCount uint64
-		for _, count := range wordMap {
-			wordCount += count
-		}
-		fmt.Fprintf(w, "Total word count: %v\n", wordCount)
-		avgCount := float64(wordCount) / entryCount
-		fmt.Fprintf(w, "Average word count: %.1f\n", avgCount)
-		fmt.Fprint(w, "\n")
 
-		if len(refFreqs) == 0 {
-			return nil // no code generation. exit early
-		}
-
-		wordStats := make([]*wordStat, len(wordMap))
-		i := 0
-		for word, count := range wordMap {
-			frequency := float64(count) / float64(wordCount)
-			var relFrequency float64
-			refFrequency := refFreqs[word]
-			if frequency > refFrequency {
-				if refFrequency > 0 {
-					relFrequency = frequency / refFrequency
-				}
-			} else {
-				relFrequency = (refFrequency / frequency) * -1
-			}
-			wordStats[i] = &wordStat{word: word, occurrences: count, frequency: relFrequency}
-			i++
-		}
-
-		sort.Slice(wordStats, func(i, j int) bool {
-			return wordStats[i].frequency > wordStats[j].frequency
-		})
-
-		topUnusualWordCount := 100
-		w := tabwriter.NewWriter(os.Stdout, 0, 0, 1, ' ', 0)
-		fmt.Fprintf(w, "Top %v unusually frequent words:\n", topUnusualWordCount)
-		for _, ws := range wordStats[:topUnusualWordCount] {
-			fmt.Fprintf(w, "%v\t%.1fX\n", ws.word, ws.frequency)
-		}
-		w.Flush()
-		fmt.Fprint(w, "\n")
-		fmt.Fprintf(w, "Top %v unusually infrequent words:\n", topUnusualWordCount)
-		for i := 1; i <= topUnusualWordCount; i++ {
-			ws := wordStats[len(wordStats)-i]
-			fmt.Fprintf(w, "%v\t%.1fX\n", ws.word, ws.frequency)
-		}
-		w.Flush()
+	if entryCount == 0 {
+		fmt.Fprint(w, "no entries found! why not try writing one with 'gurnel start'?")
+		return nil
 	}
+
+	percent := entryCount / math.Ceil(t.Sub(minDate).Hours()/24)
+	const outFormat = "Jan 2 2006"
+	fmt.Fprintf(w, "%.2f%% of days journaled since %v\n", percent*100, minDate.Format(outFormat))
+	var wordCount uint64
+	for _, count := range wordMap {
+		wordCount += count
+	}
+	fmt.Fprintf(w, "Total word count: %v\n", wordCount)
+	avgCount := float64(wordCount) / entryCount
+	fmt.Fprintf(w, "Average word count: %.1f\n", avgCount)
+	fmt.Fprint(w, "\n")
+
+	wordStats := make([]*wordStat, len(wordMap))
+	i := 0
+	for word, count := range wordMap {
+		frequency := float64(count) / float64(wordCount)
+		var relFrequency float64
+		refFrequency := refFreqs[word]
+		if frequency > refFrequency {
+			if refFrequency > 0 {
+				relFrequency = frequency / refFrequency
+			}
+		} else {
+			relFrequency = (refFrequency / frequency) * -1
+		}
+		wordStats[i] = &wordStat{word: word, occurrences: count, frequency: relFrequency}
+		i++
+	}
+
+	sort.Slice(wordStats, func(i, j int) bool {
+		return wordStats[i].frequency > wordStats[j].frequency
+	})
+
+	var topUnusualWordCount uint64
+	topUnusualWordCount = 100
+	if topUnusualWordCount > wordCount {
+		topUnusualWordCount = wordCount
+	}
+	out := tabwriter.NewWriter(w, 0, 0, 1, ' ', 0)
+	fmt.Fprintf(out, "Top %v unusually frequent words:\n", topUnusualWordCount)
+	for _, ws := range wordStats[:topUnusualWordCount] {
+		fmt.Fprintf(out, "%v\t%.1fX\n", ws.word, ws.frequency)
+	}
+	out.Flush()
+	fmt.Fprint(out, "\n")
+	fmt.Fprintf(out, "Top %v unusually infrequent words:\n", topUnusualWordCount)
+	for i := 1; i <= int(topUnusualWordCount); i++ {
+		ws := wordStats[len(wordStats)-i]
+		fmt.Fprintf(out, "%v\t%.1fX\n", ws.word, ws.frequency)
+	}
+	out.Flush()
+
 	return nil
 }
 

--- a/internal/gurnel/stats_test.go
+++ b/internal/gurnel/stats_test.go
@@ -1,0 +1,93 @@
+package gurnel
+
+import (
+	"bytes"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/mikeraimondi/gurnel/internal/test"
+)
+
+func TestStats(t *testing.T) {
+	const noentry = "NOENTRY"
+
+	testCases := []struct {
+		desc       string
+		entryWords []string
+		out        []string
+	}{
+		{
+			desc:       "with no entries",
+			entryWords: []string{},
+			out:        []string{"no entries found"},
+		},
+		{
+			desc:       "with an empty entry",
+			entryWords: []string{""},
+			out: []string{
+				"word count: 0",
+				`100.00% of days`,
+			},
+		},
+		{
+			desc:       "with one populated entry",
+			entryWords: []string{"foo bar baz"},
+			out: []string{
+				"word count: 3",
+				`100.00% of days`,
+			},
+		},
+		{
+			desc:       "with two populated entries",
+			entryWords: []string{"foo bar", "baz"},
+			out: []string{
+				"word count: 3",
+				`100.00% of days`,
+			},
+		},
+		{
+			desc:       "with one entry and one day missed",
+			entryWords: []string{noentry, "baz"},
+			out: []string{
+				"word count: 1",
+				`50.00% of days`,
+			},
+		},
+	}
+	for _, tC := range testCases {
+		t.Run(tC.desc, func(t *testing.T) {
+			dir, cleanup := test.SetupTestDir(t)
+			defer cleanup()
+			testClock := test.FixedClock{}
+
+			for i, words := range tC.entryWords {
+				if words == noentry {
+					continue
+				}
+
+				entryTime := testClock.Now().Add(-time.Duration(24*i) * time.Hour)
+				t.Log(entryTime)
+				entry, err := NewEntry(dir, entryTime)
+				if err != nil {
+					t.Fatalf("saving entry: %s", err)
+				}
+
+				defer test.WriteFile(t, entry.Path, words)()
+			}
+
+			cmd := statsCmd{}
+			out := bytes.Buffer{}
+			conf := config{clock: &testClock}
+			if err := cmd.Run(&bytes.Buffer{}, &out, []string{}, &conf); err != nil {
+				t.Fatalf("expected no error. got %s", err)
+			}
+
+			for _, expectedOut := range tC.out {
+				if !strings.Contains(strings.ToLower(out.String()), strings.ToLower(expectedOut)) {
+					t.Fatalf("expected output containing %s. got %q", expectedOut, out.String())
+				}
+			}
+		})
+	}
+}

--- a/internal/test/test.go
+++ b/internal/test/test.go
@@ -1,0 +1,48 @@
+package test
+
+import (
+	"io/ioutil"
+	"os"
+	"testing"
+	"time"
+)
+
+func SetupTestDir(t *testing.T) (dir string, cleanup func()) {
+	dir, err := ioutil.TempDir("", "gurnel_test")
+	if err != nil {
+		t.Fatalf("creating test dir: %s", err)
+	}
+	if err = os.Chdir(dir); err != nil {
+		t.Fatalf("changing to test dir: %s", err)
+	}
+
+	cleanup = func() {
+		if err := os.RemoveAll(dir); err != nil {
+			t.Fatalf("removing test dir: %s", err)
+		}
+	}
+	return dir, cleanup
+}
+
+func WriteFile(t *testing.T, path, contents string) func() {
+	f, err := os.OpenFile(path, os.O_RDWR|os.O_APPEND|os.O_SYNC, 0600)
+	if err != nil {
+		t.Fatalf("opening file: %s", err)
+	}
+
+	if _, err = f.WriteString(contents); err != nil {
+		t.Fatalf("writing file: %s", err)
+	}
+
+	return func() {
+		if err := f.Close(); err != nil {
+			t.Fatalf("closing file: %s", err)
+		}
+	}
+}
+
+type FixedClock struct{}
+
+func (c *FixedClock) Now() time.Time {
+	return time.Date(2008, time.April, 12, 16, 0, 0, 0, time.UTC)
+}

